src/processors/other/Tremolo.cpp

```diff
@@ -118,29 +121,51 @@ void Tremolo::processAudio (AudioBuffer<float>& buffer)
 {
     const auto numSamples = buffer.getNumSamples();
     waveBuffer.setSize (1, numSamples, false, false, true);
+    modBuffer.setSize (1, numSamples, false, false, true);
     phaseSmooth.setTargetValue (*rateParam * MathConstants<float>::pi / fs);
     waveSmooth.setTargetValue (*waveParam);
 
-    // fill wave buffer (-1, 1)
-    dsp::AudioBlock<float> waveBlock { waveBuffer };
-    dsp::ProcessContextReplacing<float> waveCtx { waveBlock };
-    fillWaveBuffer (waveBlock.getChannelPointer (0), numSamples, phase);
+    if (inputsConnected.contains (ModulationInput))
+    {
+        // get modulation buffer from input (-1, 1)
+        modBuffer.copyFrom (0, 0, getInputBuffer (ModulationInput), 0, 0, numSamples);
+        makeBufferMono(modBuffer);
+    }
+    else
+    {
+        // fill modulation buffer (-1, 1)
+        fillWaveBuffer (modBuffer.getWritePointer (0), numSamples, phase);
+    }
 
-    // shrink range to (0, 1)
-    waveBlock *= 0.5f;
-    waveBlock += 0.5f;
+        // copy modulation buffer to wave buffer (-1, 1)
+        waveBuffer.makeCopyOf (modBuffer, true);
+        dsp::AudioBlock<float> waveBlock { waveBuffer };
+        dsp::ProcessContextReplacing<float> waveCtx { waveBlock };
 
-    // apply depth parameter
-    auto depthVal = std::pow (depthParam->load(), 0.33f);
-    depthGainSmooth.setTargetValue (depthVal);
-    waveBlock.multiplyBy (depthGainSmooth);
-    depthAddSmooth.setTargetValue (1.0f - depthVal);
-    addSmoothed (waveBlock, depthAddSmooth);
-    filter.process (waveCtx);
+        // shrink range to (0, 1)
+        waveBlock *= 0.5f;
+        waveBlock += 0.5f;
 
-    for (int ch = 0; ch < buffer.getNumChannels(); ++ch)
+        // apply depth parameter
+        auto depthVal = std::pow (depthParam->load(), 0.33f);
+        depthGainSmooth.setTargetValue (depthVal);
+        waveBlock.multiplyBy (depthGainSmooth);
+        depthAddSmooth.setTargetValue (1.0f - depthVal);
+        addSmoothed (waveBlock, depthAddSmooth);
+        filter.process (waveCtx);
+
+        for (int ch = 0; ch < buffer.getNumChannels(); ++ch)
+        {
+            auto* x = getInputBuffer (AudioInput).getWritePointer (0);
```

Comment:
I'm not sure this is safe to call before checking `inputsConnected.contains (AudioInput)`? Also, we can probably skip a bunch of the other processing when the input is not connected, to save some CPU.

Reply:
Yeah good call, I updated this in the latest changes.